x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java

```diff
@@ -153,8 +154,15 @@ protected void taskOperation(StopDataFrameTransformAction.Request request, DataF
                 return;
             }
 
-            transformTask.stop();
-            listener.onResponse(new StopDataFrameTransformAction.Response(Boolean.TRUE));
+            // To cover strange state race conditions, we adjust the variable first (which writes to cluster state if it is different)
+            // then we stop the transform
+            transformTask.setShouldStopAtCheckpoint(request.isWaitForCheckpoint(), ActionListener.wrap(
```

Comment:
Since the default value for `wait_for_checkpoint` is `true` and folks could call this against `_all` transforms, this could get pretty expensive as there will be a cluster state update for each individual task.

Do we want to reconsider the default value of `true` or do we think this behavior is acceptable?

Reply:
The difficulty with having it default to `true` within the task, is now we will need two booleans:

* one that indicates that when we stop, we should wait for the checkpoint
* one that indicates that we are stopping

Because checkpoints are completed when `onFinish` is called, we need a way to determine that we are "stopping" and the task should complete. In this PR, we are relying on the flag being set to indicate that the task should stop when `onFinish` is completed.